### ISSUE TITLE
Line numbers in error messages are wrong

### ISSUE THREAD

**Issue by bardo84** [marcobambini/gravity#389]
I introduced different syntax errors in a running code.
The resulting error messages are always larger than the line in question.
Subsequent detected errors did not conform to equally offset numbers.

How to debug a Gravity script?

**Comment by marcobambini** [comment]
Hi @bardo84, can you please post an example?

**Comment by mwasplund** [comment]
Repro steps
----------
Test.gravity
```swift
func main() {
	System.print("Run Tests!")
	error
}
```
Run
```
gravity Test.gravity
```

Expected
-------------
Error reported in file "Test.gravity" on line 3 character 2

Actual
-----------
```
SEMANTIC ERROR on 0 (5,2): Identifier error not found.
```

**Comment by mwasplund** [comment]
Should the '0' value be the file name?

**Comment by mwasplund** [comment]
FYI: Windows 11 running VS 2022 Version 17.4.3

**Comment by marcobambini** [comment]
Thanks @mwasplund, I see that you closed your pull request? That means that it was not ready to be applied?

**Comment by marcobambini** [comment]
> Should the '0' value be the file name?

'0' is the fileID, the user is responsible to save the filename associated to each fileID.
See comment in https://github.com/marcobambini/gravity/blob/971337cc01877d62972e42cab5099379c5b97f12/src/cli/gravity.c#L64 to get more information about to use the fileID.

**Comment by mwasplund** [comment]
Yes, it appears that in some cases when calling is_newline the current offset is pointing to the actual current token and in some cases it is already pointing to next. Still working out why this is the case.

**Comment by marcobambini** [comment]
I think that a good starting point to investigate is the `is_newline` function in gravity_lexer.c
https://github.com/marcobambini/gravity/blob/971337cc01877d62972e42cab5099379c5b97f12/src/compiler/gravity_lexer.c#L76

**Comment by mwasplund** [comment]
That was were I was looking, I had updated the checks in that method to use PEEK_NEXT instead of PEEK_CURRENT. This resolved a lot of issues where the LF was getting missed and then double counted. However, it appears that in some cases the caller is "peeking" in which case the next char is actually current... I sent an update to pass in the exact next and next2 characters to the is_newline call. Not the prettiest, but it is working. Let me know if you see a cleaner way to achieve the difference in state between calls.

**Comment by marcobambini** [comment]
Fixed in daefac8.

The reported row drifted on sources saved with CR+LF line endings, which matches "always larger than the line in question" and the offsets not being consistent between one error and the next: the drift grows by one for every line read so far.

`is_newline()` in the lexer used `PEEK_CURRENT` to look at the character following `c`. That only holds where the caller had already consumed `c` (the comment scanner and `gravity_lexer_skip_line`). In `gravity_lexer_next`, `c` was still the character sitting at the current offset, so the CR of a CR+LF pair never saw the LF next to it, the pair was taken for two separate line breaks, and the counter ran ahead by one per line. A file with an error on row 5 reported it on row 9.

The lookahead is now passed in explicitly at each call site, `newline_length()` returns the size of the terminator without consuming anything, and the callers that still hold the offset skip the whole sequence in one go. Files with LF endings are unaffected — they were already correct.

There is a regression test at `test/unittest/bugfix_crlf_lineno.gravity`, stored with CR+LF endings on purpose, asserting `error_row` and `error_col`.

Thanks for the report, and apologies it took this long. It was found and diagnosed correctly in #401 by @mwasplund — the fix that went in is that patch with the string and number scanners corrected.
